src/merge_driver.c

```diff
@@ -0,0 +1,397 @@
+	GIT_UNUSED(src);
+
+	*payload = (void *)GIT_MERGE_FILE_FAVOR_NORMAL;
+	return 0;
+}
+
+static int merge_driver_union_check(
+	git_merge_driver *self,
+	void **payload,
+	const char *name,
+	const git_merge_driver_source *src)
+{
+	GIT_UNUSED(self);
+	GIT_UNUSED(name);
+	GIT_UNUSED(src);
+
+	*payload = (void *)GIT_MERGE_FILE_FAVOR_UNION;
+	return 0;
+}
+
+static int merge_driver_binary_apply(
+	git_merge_driver *self,
+	void **payload,
+	const char **path_out,
+	uint32_t *mode_out,
+	git_buf *merged_out,
+	const git_merge_driver_source *src)
+{
+	GIT_UNUSED(self);
+	GIT_UNUSED(payload);
+	GIT_UNUSED(path_out);
+	GIT_UNUSED(mode_out);
+	GIT_UNUSED(merged_out);
+	GIT_UNUSED(src);
+
+	return GIT_EMERGECONFLICT;
+}
+
+static int merge_driver_entry_cmp(const void *a, const void *b)
+{
+	const git_merge_driver_entry *entry_a = a;
+	const git_merge_driver_entry *entry_b = b;
+
+	return strcmp(entry_a->name, entry_b->name);
+}
+
+static int merge_driver_entry_search(const void *a, const void *b)
+{
+	const char *name_a = a;
+	const git_merge_driver_entry *entry_b = b;
+
+	return strcmp(name_a, entry_b->name);
+}
+
+static void merge_driver_registry_shutdown(void)
+{
+	struct merge_driver_registry *reg;
+	git_merge_driver_entry *entry;
+    size_t i;
+
+    if ((reg = git__swap(merge_driver_registry, NULL)) == NULL)
+        return;
+
+    git_vector_foreach(&reg->drivers, i, entry) {
+        if (entry && entry->driver->shutdown)
+            entry->driver->shutdown(entry->driver);
+
+        git__free(entry);
+    }
+
+    git_vector_free(&reg->drivers);
+    git__free(reg);
+}
+
+git_merge_driver git_merge_driver__normal = {
+	GIT_MERGE_DRIVER_VERSION,
+	NULL,
+	NULL,
+	NULL,
+	merge_driver_apply
+};
+
+git_merge_driver git_merge_driver__text = {
+	GIT_MERGE_DRIVER_VERSION,
+	NULL,
+	NULL,
+	merge_driver_text_check,
+	merge_driver_apply
+};
+
+git_merge_driver git_merge_driver__union = {
+	GIT_MERGE_DRIVER_VERSION,
+	NULL,
+	NULL,
+	merge_driver_union_check,
+	merge_driver_apply
+};
+
+git_merge_driver git_merge_driver__binary = {
+	GIT_MERGE_DRIVER_VERSION,
+	NULL,
+	NULL,
+	NULL,
+	merge_driver_binary_apply
+};
+
+static int merge_driver_registry_initialize(void)
+{
+	struct merge_driver_registry *reg;
+	int error = 0;
+
+	if (merge_driver_registry)
+		return 0;
+
+	reg = git__calloc(1, sizeof(struct merge_driver_registry));
+	GITERR_CHECK_ALLOC(reg);
+
+    if ((error = git_vector_init(&reg->drivers, 3, merge_driver_entry_cmp)) < 0)
+		goto done;
+	
+    reg = git__compare_and_swap(&merge_driver_registry, NULL, reg);
+
+    if (reg != NULL)
+        goto done;
+
+    git__on_shutdown(merge_driver_registry_shutdown);
+
+	if ((error = git_merge_driver_register(
+			merge_driver_name__text, &git_merge_driver__text)) < 0 ||
+		(error = git_merge_driver_register(
+			merge_driver_name__union, &git_merge_driver__union)) < 0 ||
+		(error = git_merge_driver_register(
+			merge_driver_name__binary, &git_merge_driver__binary)) < 0)
+		goto done;
```

Comment:
These registrations happen after we've set the global pointer. If we're pre-empted just before this, a second thread could come, see `merge_driver_registry` is set and return immediately. If we don't come back in time, the second thread might try to look up in the registration which we haven't fully initialized.